corehq/ex-submodules/pillowtop/run_pillowtop.py

```diff
@@ -17,24 +24,46 @@ def start_pillows(pillows=None):
     """
     run_pillows = pillows or get_all_pillow_instances()
 
-    try:
-        while True:
-            jobs = []
-            print("[pillowtop] Starting pillow processes")
-            for pillow_class in run_pillows:
-                p = multiprocessing.Process(target=pillow_class.run)
-                p.start()
-                jobs.append(p)
-            print("[pillowtop] all processes started, pids: %s" % ([x.pid for x in jobs]))
-            for j in jobs:
-                j.join()
-            print("[pillowtop] All processes complete, restarting")
-    except KeyboardInterrupt:
-        sys.exit()
+    while True:
+        pids = []
+        handler = functools.partial(_remove_prometheus_metric_files, pids)
+        with SignalHandlerContext([signal.SIGINT, signal.SIGTERM, signal.SIGHUP], handler):
+            try:
+                print("[pillowtop] Starting pillow processes")
+                jobs = []
+                for pillow_class in run_pillows:
+                    p = multiprocessing.Process(target=pillow_class.run)
+                    p.start()
+                    jobs.append(p)
+
+                pids.extend(x.pid for x in jobs)
+                print(f"[pillowtop] all processes started, pids: {pids}")
+
+                for j in jobs:
+                    j.join()
+            finally:
+                handler()
+
+        print("[pillowtop] All processes complete, restarting")
 
 
 def start_pillow(pillow_instance):
-    while True:
-        print("Starting pillow %s.run()" % pillow_instance.__class__)
-        pillow_instance.run()
-        print("Pillow %s.run() completed, restarting" % pillow_instance.__class__)
+    handler = functools.partial(_remove_prometheus_metric_files, [os.getpid()])
+    with SignalHandlerContext([signal.SIGINT, signal.SIGTERM, signal.SIGHUP], handler):
+        try:
+            while True:
+                print("Starting pillow %s.run()" % pillow_instance.__class__)
+                pillow_instance.run()
+                print("Pillow %s.run() completed, restarting" % pillow_instance.__class__)
+        finally:
+            handler()
+
+
+def _remove_prometheus_metric_files(pids):
+    path = os.environ.get('prometheus_multiproc_dir')
+    if not path:
+        return
+
+    for pid in pids:
+        for file in glob.glob(os.path.join(path, '*_{}.db'.format(pid))):
+            os.remove(file)
```

Comment:
✅ Looks like this follows the same pattern that ended up fixing the error in `gunicorn_conf.py`: https://github.com/dimagi/commcare-hq/blob/a00aa17609449e009fbc71235402cc7c7ebe5086/deployment/gunicorn/gunicorn_conf.py#L49-L50. If no files match then `glob.glob` won't find anything, so it'll never call `os.remove` on a file path that doesn't exist.

Just saying my thoughts aloud here, looks good to me